File: packages-js/bpmn-flowaudit/src/neutralize/neutralize.ts
/**
 * Neutralising: export without references to bodies, persons and findings.
 *
 * Same behaviour as `auditcore_bpmn` neutralisation, on the XML DOM:
 * - display names of pools and lanes → role label (actor or profile
 *   alias; otherwise „Stelle n“); the same names are replaced in task
 *   prefixes, labels and documentation;
 * - replacements named by the application (e.g. names of authorities);
 * - e-mail addresses, IBAN, amounts, company names, salutations with names
 *   and identifiers (SAP, MaStR, file numbers, funding codes);
 * - internal notes, audit steps, findings, finding references, finding
 *   markers and finding colours, internal sources and everything marked
 *   `vertraulich="true"`;
 * - person fields (`autor`, `freigegebenDurch`, `pruefer`).
 *
 * Automatic recognition does not replace a review: the report lists every
 * change so a person can check the result.
 */

import { FLOWAUDIT_NAMESPACE } from '../schema/descriptor'
import { attributeName, elementsByNs, parseXml, serializeXml } from '../model/xmlDom'
import { roleFromText, roleOf, type ProfileData } from '../profile/profile'
import { label } from '../schema/vocabulary'
import {
  CODE_ATTRIBUTES,
  FINDING_COLORS,
  FINDING_MARKERS,
  INTERNAL_SOURCES,
  PERSON_ATTRIBUTES,
  REMOVED_ELEMENTS,
  TEXT_PATTERNS,
  type ReplacementKind,
} from './patterns'

const BPMN = 'http://www.omg.org/spec/BPMN/20100524/MODEL'
const BPMNDI = 'http://www.omg.org/spec/BPMN/20100524/DI'
const COLOR_ATTRIBUTES = [
  ['http://bpmn.io/schema/bpmn/biocolor/1.0', 'fill'],
  ['http://bpmn.io/schema/bpmn/biocolor/1.0', 'stroke'],
  ['http://www.omg.org/spec/BPMN/non-normative/color/1.0', 'background-color'],
  ['http://www.omg.org/spec/BPMN/non-normative/color/1.0', 'border-color'],
] as const

export interface Replacement {
  elementId: string | null
  location: string
  kind: ReplacementKind
  replacement: string
  original?: string
}

export interface NeutralizationResult {
  xml: string
  replacements: Replacement[]
  /** Count per kind. */
  summary: Record<string, number>
}

export interface NeutralizeOptions {
  profile?: ProfileData | null
  /** Names known to the application → replacement. */
  replacements?: Record<string, string>
  /** Include the original texts in the report (internal review only). */
  withOriginals?: boolean
}

function isConfidential(element: Element): boolean {
  return (element.getAttribute('vertraulich') ?? '').toLowerCase() === 'true'
}

function shouldRemove(element: Element): boolean {
  const name = element.localName
  const type = element.getAttribute('typ') ?? ''
  const kind = element.getAttribute('art') ?? ''
  return (
    REMOVED_ELEMENTS.has(name) ||
    isConfidential(element) ||
    (name === 'kennzeichen' && FINDING_MARKERS.has(type)) ||
    (name === 'quelle' && INTERNAL_SOURCES.has(kind)) ||
    (name === 'verweis' && kind === 'feststellung_ref')
  )
}

function ownerId(node: Element): string | null {
  let current: Element | null = node
  while (current && !current.getAttribute('id')) current = current.parentElement
  return current?.getAttribute('id') ?? null
}

class Neutralizer {
  readonly report: Replacement[] = []
  private readonly names = new Map<string, string>()
  private counter = 0

  constructor(
    private readonly document: XMLDocument,
    private readonly options: NeutralizeOptions,
  ) {
    for (const [old, replacement] of Object.entries(options.replacements ?? {})) if (old.trim()) this.names.set(old.trim(), replacement)
  }

  log(elementId: string | null, location: string, kind: ReplacementKind, replacement: string, original?: string): void {
    this.report.push({ elementId, location, kind, replacement, ...(this.options.withOriginals && original ? { original } : {}) })
  }

  private actorOf(element: Element): Element | undefined {
    const container = Array.from(element.children).find((child) => child.namespaceURI === BPMN && child.localName === 'extensionElements')
    return Array.from(container?.children ?? []).find((child) => child.namespaceURI === FLOWAUDIT_NAMESPACE && child.localName === 'akteur')
  }

  roles(): void {
    const containers = [...elementsByNs(this.document, BPMN, 'participant'), ...elementsByNs(this.document, BPMN, 'lane')]
    for (const element of containers) {
      const old = (element.getAttribute('name') ?? '').trim()
      const actor = this.actorOf(element)
      const code = actor?.getAttribute('rolle') || roleFromText(this.options.profile, old)
      const role = roleOf(this.options.profile, code)
      const replacement = role ? label(role.label) : `Stelle ${(this.counter += 1)}`
      const displayName = actor?.getAttribute('anzeigename')
      if (displayName && !this.names.has(displayName)) this.names.set(displayName, replacement)
      if (old && old !== replacement) {
        if (!this.names.has(old)) this.names.set(old, replacement)
        element.setAttribute('name', replacement)
        this.log(element.getAttribute('id'), 'name', 'anzeigename', replacement, old)
      }
    }
  }

  clean(value: string, elementId: string | null, location: string): string {
    let result = value
    for (const old of [...this.names.keys()].sort((a, b) => b.length - a.length)) {
      if (old && result.includes(old)) {
        result = result.split(old).join(this.names.get(old) as string)
        this.log(elementId, location, 'anzeigename', this.names.get(old) as string, old)
      }
    }
    for (const { kind, pattern, replacement } of TEXT_PATTERNS) {
      result = result.replace(pattern, (match) => {
        this.log(elementId, location, kind, replacement, match)
        return replacement
      })
    }
    return result === value ? result : result.replace(/[ \t]{2,}/g, ' ').replace(/ ,/g, ',').replace(/ \./g, '.')
  }

  flowaudit(): void {
    for (const container of elementsByNs(this.document, BPMN, 'extensionElements')) {
      const elementId = ownerId(container)
      for (const child of Array.from(container.children)) {
        if (child.namespaceURI !== FLOWAUDIT_NAMESPACE) continue
        this.flowauditChild(container, child, elementId)
      }
      if (container.children.length === 0) container.parentNode?.removeChild(container)
    }
  }

  private flowauditChild(container: Element, child: Element, elementId: string | null): void {
    const name = child.localName
    if (shouldRemove(child)) {
      container.removeChild(child)
      this.log(elementId, `flowaudit:${name}`, 'entfernt', '')
      return
    }
    if (name === 'akteur' && child.getAttribute('anzeigename')) {
      child.removeAttribute('anzeigename')
      this.log(elementId, 'flowaudit:akteur/@anzeigename', 'anzeigename', '')
    }
    if (name === 'diagrammInfo') this.diagramInfo(child, elementId)
  }

  private diagramInfo(info: Element, elementId: string | null): void {
    for (const attribute of PERSON_ATTRIBUTES) {
      if (info.hasAttribute(attribute)) {
        info.removeAttribute(attribute)
        this.log(elementId, `diagrammInfo/@${attribute}`, 'person', '')
      }
    }
    for (const child of Array.from(info.children)) {
      if (child.localName === 'feststellung' || shouldRemove(child)) {
        info.removeChild(child)
        this.log(elementId, `diagrammInfo/flowaudit:${child.localName}`, 'entfernt', '')
      }
    }
  }

  private cleanNode(node: Element): void {
    const id = ownerId(node)
    const inFlowaudit = node.namespaceURI === FLOWAUDIT_NAMESPACE
    for (const attribute of Array.from(node.attributes)) {
      if (attribute.namespaceURI || attribute.name.includes(':')) continue
      const relevant = attribute.name === 'name' || (inFlowaudit && !CODE_ATTRIBUTES.has(attribute.name))
      if (!relevant) continue
      if (inFlowaudit && attribute.name === 'pruefer') continue
      const cleaned = this.clean(attribute.value, id, `@${attribute.name}`)
      if (cleaned !== attribute.value) node.setAttribute(attribute.name, cleaned)
    }
    const text = node.textContent ?? ''
    if (node.children.length === 0 && text.trim()) {
      const cleaned = this.clean(text, id, node.localName)
      if (cleaned !== text) node.textContent = cleaned
    }
  }

  private colors(shape: Element): void {
    for (const attribute of Array.from(shape.attributes)) {
      const { namespace: ns, local } = attributeName(attribute, shape)
      const isColor = COLOR_ATTRIBUTES.some(([namespace, name]) => ns === namespace && local === name)
      if (isColor && FINDING_COLORS.has(attribute.value.toLowerCase())) {
        shape.removeAttributeNode(attribute)
        this.log(shape.getAttribute('bpmnElement'), 'DI-Farbe', 'befundfarbe', '')
      }
    }
  }

  texts(): void {
    for (const node of Array.from(this.document.getElementsByTagName('*'))) {
      if (node.namespaceURI === BPMN || node.namespaceURI === FLOWAUDIT_NAMESPACE) this.cleanNode(node)
      if (node.namespaceURI === BPMNDI && node.localName === 'BPMNShape') this.colors(node)
    }
  }
}

/** Neutralises a diagram; `replacements` adds names known to the application. */
export function neutralize(xml: string, options: NeutralizeOptions = {}): NeutralizationResult {
  const document = parseXml(xml)
  if (!document) throw new Error('Das BPMN-XML ist nicht wohlgeformt.')
  const worker = new Neutralizer(document, options)
  worker.roles()
  worker.flowaudit()
  worker.texts()
  const summary: Record<string, number> = {}
  for (const item of worker.report) summary[item.kind] = (summary[item.kind] ?? 0) + 1
  return { xml: serializeXml(document), replacements: worker.report, summary }
}

File: packages/auditcore_bpmn/src/auditcore_bpmn/neutralize.py
"""Neutralisieren: Export ohne Stellen-, Personen- und Befundbezug.

:func:`neutralize` liefert das neutralisierte XML und einen Bericht, was
ersetzt oder entfernt wurde:

* Anzeigenamen von Lanes (und Pools mit Rolle) → Rollenbezeichnung (Akteur
  oder Rollen-Alias des Profils, sonst „Stelle n“). Ganze Lane-Namen werden
  in Beschriftungen und Aufgabenpräfixen („Name: …“) ersetzt; Anzeigenamen
  der Akteure und Namensteile zusammengesetzter Lane-Namen, die keine Rolle
  bezeichnen (z. B. „Beispielbank“ in „Beispielbank (ZGS)“), überall;
* von der Anwendung genannte Ersetzungen (z. B. Behördennamen) überall;
* E-Mail-Adressen, IBAN, Geldbeträge, Unternehmensnamen (Rechtsformzusatz),
  Anreden mit Namen und Kennnummern (SAP, MaStR, Aktenzeichen, Förderkennzeichen);
* interne Notizen, Prüfschritte, Feststellungen und Feststellungsbezüge
  (auch im Text, z. B. „Feststellung T15 F1“, „(F3)“), befundbezogene
  Kennzeichen und Rotfärbungen, Quellen aus Interviews und Arbeitspapieren
  sowie alle mit ``vertraulich="true"`` markierten Angaben;
* Personenfelder (``autor``, ``freigegebenDurch``, ``pruefer``).

Automatische Erkennung ersetzt keine Durchsicht; der Bericht nennt jede Stelle.
"""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass, field
from xml.etree import ElementTree as ET

from .model import BpmnDocument
from .namespaces import BIOC_NS, BPMN_NS, BPMNDI_NS, COLOR_NS, FLOWAUDIT_NAMESPACE, local_name, namespace_of, q
from .profiles import Profile, load_profile
from .serialize import serialize
from .vocabulary import label
from .writer import copy_document

FINDING_MARKERS = frozenset(
    {
        "feststellung",
        "feststellung_formell",
        "feststellung_finanziell",
        "offener_nachweis",
        "ohne_befund",
        "soll_ohne_regelung",
    }
)
FINDING_COLORS = frozenset({"#fce8e6", "#b3261e", "#ffcdd2", "#b71c1c", "#ffe0e0", "#cc0000"})
INTERNAL_SOURCES = frozenset({"interview", "arbeitspapier", "durchlauftest"})
REMOVED_ELEMENTS = frozenset({"interneNotiz", "notiz", "pruefschritt", "feststellung"})
PERSON_ATTRIBUTES = ("autor", "freigegebenDurch", "pruefer")
#: FlowAudit-Attribute mit Codes oder Verweisen (werden nicht als Text bereinigt).
_CODE_ATTRIBUTES = frozenset(
    {
        "id",
        "typ",
        "art",
        "ka",
        "bk",
        "rolle",
        "status",
        "kontrollen",
        "kontrolle",
        "schemaVersion",
        "profil",
        "ergebnis",
        "einheit",
        "kategorie",
        "inhaerent",
        "kontrollrisiko",
        "restrisiko",
        "einstufung",
        "durchfuehrung",
        "vertraulich",
        "vertraulichkeit",
        "variante",
        "foerderperiode",
        "celex",
        "eli",
        "url",
        "gueltigAb",
        "gueltigBis",
        "freigegebenAm",
        "vksStichtag",
        "schluesselkontrolle",
    }
)
_DI_COLOR_ATTRIBUTES = (
    q(BIOC_NS, "fill"),
    q(BIOC_NS, "stroke"),
    q(COLOR_NS, "background-color"),
    q(COLOR_NS, "border-color"),
)

PATTERNS: tuple[tuple[str, re.Pattern[str], str], ...] = (
    ("email", re.compile(r"[\w.+-]+@[\w-]+(?:\.[\w-]+)+"), "[E-Mail entfernt]"),
    ("iban", re.compile(r"\b[A-Z]{2}\d{2}(?: ?[A-Z0-9]{4}){3,7}(?: ?[A-Z0-9]{1,3})?\b"), "[IBAN entfernt]"),
    (
        "kennnummer",
        re.compile(
            r"\b(?:SAP|MaStR|Az\.|Aktenzeichen|Förderkennzeichen|FKZ|Projektnummer|Vorhabensnummer|Vorhabennummer)"
            r"(?:-Nr\.|-Nummer| Nr\.| Nummer)?[\s.:]*[A-Z0-9][A-Z0-9./-]{3,}"
        ),
        "[Kennnummer entfernt]",
    ),
    ("kennnummer", re.compile(r"\b(?:SEE|SME|ABR|SNB|EEG)\d{9,}\b"), "[Kennnummer entfernt]"),
    (
        "betrag",
        re.compile(
            r"(?:(?:€|EUR)\s?\d{1,3}(?:[.\s]\d{3})*(?:,\d{1,2})?(?:\s?(?:Mio\.|Mrd\.|Tsd\.))?)"
            r"|(?:\d{1,3}(?:[.\s]\d{3})*(?:,\d{1,2})?|\d+(?:,\d{1,2})?)\s?(?:Mio\.\s?|Mrd\.\s?|Tsd\.\s?)?(?:€|EUR|Euro)\b"
        ),
        "[Betrag entfernt]",
    ),
    (
        "unternehmen",
        re.compile(
            r"(?:[A-ZÄÖÜ][\wÄÖÜäöüß&.'-]*\s){0,5}[A-ZÄÖÜ][\wÄÖÜäöüß&.'-]*\s"
            r"(?:GmbH & Co\. KG|gGmbH|GmbH|mbH|AG|KG|OHG|e\.\s?V\.|eG|GbR|SE|UG(?: \(haftungsbeschränkt\))?|Ltd\.?)"
            r"(?=[\s,.;:)\]]|$)"
        ),
        "[Unternehmen]",
    ),
    (
        "person",
        re.compile(r"\b(?:Herr|Frau|Hr\.|Fr\.)\s+(?:Dr\.\s+|Prof\.\s+)?[A-ZÄÖÜ][a-zäöüß]+(?:-[A-ZÄÖÜ][a-zäöüß]+)?"),
        "[Person]",
    ),
    (
        "feststellungsbezug",
        re.compile(
            r"(?:Feststellung(?:en)?|Themenvermerk(?:e)?)\s+T\d+(?:\s+F\d+)?"
            r"(?:(?:\s*,\s*|\s+und\s+)(?:T\d+\s+)?F?\d+)*"
        ),
        "",
    ),
    ("feststellungsbezug", re.compile(r"\s?\((?:F\d+(?:,\s*)?)+\)"), ""),
)


@dataclass(frozen=True)
class Replacement:
    """Eine Ersetzung oder Entfernung (Original nur auf Wunsch)."""

    element_id: str | None
    location: str
    kind: str
    new: str
    old: str | None = None


@dataclass
class NeutralizationResult:
    """Neutralisiertes XML und Bericht."""

    xml: str
    replacements: list[Replacement] = field(default_factory=list)

    def summary(self) -> dict[str, int]:
        """Anzahl der Ersetzungen je Art."""
        counts: dict[str, int] = {}
        for item in self.replacements:
            counts[item.kind] = counts.get(item.kind, 0) + 1
        return dict(sorted(counts.items()))

    def to_dict(self) -> dict[str, object]:
        """JSON-fähige Darstellung."""
        return {
            "summary": self.summary(),
            "replacements": [{k: v for k, v in vars(r).items() if v is not None} for r in self.replacements],
        }


def _removable(name: str, child: ET.Element) -> bool:
    return (
        name in REMOVED_ELEMENTS
        or (child.get("vertraulich") or "").lower() == "true"
        or (name == "kennzeichen" and child.get("typ") in FINDING_MARKERS)
        or (name == "quelle" and child.get("art") in INTERNAL_SOURCES)
        or (name == "verweis" and child.get("art") == "feststellung_ref")
    )


class _Neutralizer:
    def __init__(
        self, document: BpmnDocument, profile: Profile, replacements: Mapping[str, str], keep_originals: bool
    ) -> None:
        self.document = document
        self.profile = profile
        self.names = {old.strip(): new for old, new in replacements.items() if old.strip()}
        self.labels: dict[str, str] = {}
        self.report: list[Replacement] = []
        self.keep_originals = keep_originals
        self.counter = 0

    def log(self, element_id: str | None, location: str, kind: str, new: str, old: str | None = None) -> None:
        """Hält eine Ersetzung fest."""
        self.report.append(Replacement(element_id, location, kind, new, old if self.keep_originals else None))

    def role_name(self, name: str, role_code: str | None, *, numbered: bool) -> str | None:
        """Rollenbezeichnung einer Stelle oder „Stelle n“ (nur Lanes)."""
        role = self.profile.role(role_code or self.profile.role_from_text(name))
        if role is not None:
            return label(role.labels)
        if not numbered:
            return None
        self.counter += 1
        return f"Stelle {self.counter}"

    def _name_parts(self, name: str, new: str) -> None:
        parts = [p.strip() for p in re.split(r"[()–—,/]|\s-\s", name) if p.strip()]
        if len(parts) < 2:
            return
        for part in parts:
            if len(part) >= 3 and self.profile.role_from_text(part) is None:
                self.names.setdefault(part, new)

    def rename_bodies(self) -> None:
        """Ersetzt Namen von Lanes und Pools mit Rolle."""
        for element in (*self.document.participants, *self.document.lanes):
            old = (element.name or "").strip()
            actor = element.extensions.actor
            new = self.role_name(old, actor.role if actor else None, numbered=element.type == "lane")
            if new is None:
                continue
            if actor and actor.display_name:
                self.names.setdefault(actor.display_name, new)
            if old and old != new:
                self.labels.setdefault(old, new)
                self._name_parts(old, new)
                self.document.xml_element(element.id).set("name", new)
                self.log(element.id, "name", "anzeigename", new, old)

    def _replace_labels(self, value: str, element_id: str | None, location: str) -> str:
        stripped = value.strip()
        for old, new in self.labels.items():
            if stripped == old:
                self.log(element_id, location, "anzeigename", new, old)
                return new
            if stripped.startswith(old + ":"):
                self.log(element_id, location, "anzeigename", new, old)
                return new + stripped[len(old) :]
        return value

    def clean(self, value: str, element_id: str | None, location: str) -> str:
        """Bereinigt einen Text (Namen, Muster)."""
        result = self._replace_labels(value, element_id, location)
        for old in sorted(self.names, key=len, reverse=True):
            if old in result:
                result = result.replace(old, self.names[old])
                self.log(element_id, location, "anzeigename", self.names[old], old)
        for kind, pattern, replacement in PATTERNS:
            for match in pattern.finditer(result):
                self.log(element_id, location, kind, replacement, match.group(0))
            result = pattern.sub(replacement, result)
        if result != value:
            result = re.sub(r"[ \t]{2,}", " ", result).replace(" ,", ",").replace(" .", ".")
        return result

    def strip_extensions(self, node: ET.Element, element_id: str) -> None:
        """Entfernt vertrauliche und befundbezogene FlowAudit-Angaben."""
        container = node.find(q(BPMN_NS, "extensionElements"))
        if container is None:
            return
        for child in list(container):
            if namespace_of(child.tag) != FLOWAUDIT_NAMESPACE:
                continue
            name = local_name(child.tag)
            if _removable(name, child):
                container.remove(child)
                self.log(element_id, f"flowaudit:{name}", "entfernt", "")
            elif name == "akteur":
                self.remove_attributes(child, element_id, ("anzeigename",), "anzeigename")
            elif name == "diagrammInfo":
                self.strip_info(child, element_id)
        if len(container) == 0:
            node.remove(container)

    def remove_attributes(self, node: ET.Element, element_id: str | None, names: tuple[str, ...], kind: str) -> None:
        """Entfernt Attribute und hält es fest."""
        for attribute in names:
            if attribute in node.attrib:
                del node.attrib[attribute]
                self.log(element_id, f"@{attribute}", kind, "")

    def strip_info(self, info: ET.Element, element_id: str) -> None:
        """Bereinigt die Diagramm-Infos."""
        self.remove_attributes(info, element_id, PERSON_ATTRIBUTES, "person")
        for child in list(info):
            name = local_name(child.tag)
            if _removable(name, child):
                info.remove(child)
                self.log(element_id, f"diagrammInfo/flowaudit:{name}", "entfernt", "")

    def clean_node(self, node: ET.Element, owner_id: str | None) -> None:
        """Bereinigt Texte und Textattribute eines Knotens."""
        space = namespace_of(node.tag)
        if space not in (BPMN_NS, FLOWAUDIT_NAMESPACE):
            return
        self.remove_attributes(node, owner_id, ("pruefer",) if local_name(node.tag) == "pruefschritt" else (), "person")
        for attribute, value in list(node.attrib.items()):
            textual = attribute == "name" or (space == FLOWAUDIT_NAMESPACE and attribute not in _CODE_ATTRIBUTES)
            if textual:
                node.set(attribute, self.clean(value, owner_id, f"@{attribute}"))
        if node.text and node.text.strip() and len(node) == 0:
            node.text = self.clean(node.text, owner_id, local_name(node.tag))

    def clean_colors(self, node: ET.Element) -> None:
        """Entfernt befundbezogene Farben einer DI-Form."""
        for attribute in _DI_COLOR_ATTRIBUTES:
            if (node.get(attribute) or "").lower() in FINDING_COLORS:
                del node.attrib[attribute]
                self.log(node.get("bpmnElement"), "DI-Farbe", "befundfarbe", "")

    def clean_texts(self) -> None:
        """Bereinigt alle Texte des Dokuments."""
        owners: dict[ET.Element, str | None] = {}
        for parent in self.document.root.iter():
            owner = parent.get("id") or owners.get(parent)
            for child in parent:
                owners[child] = child.get("id") or owner
        for node in self.document.root.iter():
            if not isinstance(node.tag, str):
                continue
            if node.tag == q(BPMNDI_NS, "BPMNShape"):
                self.clean_colors(node)
            else:
                self.clean_node(node, owners.get(node, node.get("id")))

    def run(self) -> None:
        """Führt alle Schritte aus."""
        self.rename_bodies()
        for element in list(self.document):
            self.strip_extensions(self.document.xml_element(element.id), element.id)
        self.clean_texts()


def neutralize(
    source: str | bytes | BpmnDocument,
    *,
    profile: Profile | None = None,
    replacements: Mapping[str, str] | None = None,
    keep_originals: bool = False,
) -> NeutralizationResult:
    """Neutralisiert ein Diagramm; ``replacements`` ergänzt Namen, die die Anwendung kennt.

    ``keep_originals`` nimmt die ersetzten Originaltexte in den Bericht auf
    (nur für die interne Durchsicht, nicht zur Weitergabe).
    """
    document = copy_document(source)
    worker = _Neutralizer(document, profile or load_profile(), replacements or {}, keep_originals)
    worker.run()
    return NeutralizationResult(serialize(document.parsed), worker.report)
